### ISSUE TITLE
Taproot::control_block shouldn't make you move a whole `ScriptBuf` into a tuple

### ISSUE THREAD

**Issue by apoelstra** [rust-bitcoin/rust-bitcoin#3603]
Not sure if it's related here, but I have an issue with the [`control_block`](https://github.com/rust-bitcoin/rust-bitcoin/blob/73e33e58089b23bf8f46ef6d54f40972d9e2d333/bitcoin/src/taproot/mod.rs#L291) method. As it accepts `script_ver: &(ScriptBuf, LeafVersion)` the caller must move the script into the reference of the tuple which is not the same as a tuple of references (e.g. `(&ScriptBuf, &LeafVersion)`). 

Usually, this could be neglected as scripts are relatively small, but in the BitVM, script sizes are up to 300-400 kilobytes each, and TapTree can consist of thousands of such scripts, which, during creation of control blocks for each of them, would require thousands of allocations of 300-400 kilobytes chunks.

The simplest example is:

```rust
let UNSPENDABLE_KEY: XOnlyPublicKey = 
    "50929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0"
        .parse()
        .unwrap();
let ctx = Secp256k1::new();

let script = some_very_large_script();
let spending_info = TaprootBuilder::new()
    .add_leaf(1, /* First move */ script.clone())
    .unwrap()
    .finalize(&ctx, *UNSPENDABLE_KEY)
    .unwrap();

let control_block = spending_info.control_block(&(/* second one */ script.clone(), LeafVersion::TapScript));

// Then something is done with a script later
println!("{}", script.to_asm_string());
```

Maybe, I'm using the API in a wrong way or missing a discussion where the team considered this solution optimal. Also, I haven't done any benchmarks or profiling yet as the current code is in the early development stage, but felt that you may find it a little bit frustrating too

_Originally posted by @Velnbur in https://github.com/rust-bitcoin/rust-bitcoin/issues/1614#issuecomment-2470017156_

**Comment by apoelstra** [comment]
We should drop the tuple and use `Script` instead of `ScriptBuf`.

This original API was intended for use with the `TaprootSpendInfo::script_map` API which maintains a btreemap from scripts to control blocks. But this API is pretty weird and special-case (e.g. it doesn't make sense for rust-miniscript, and we are going to replace our use of `TaprootSpendInfo` everywhere with something more appropriate) and I'm not aware of anybody using Taproot this way.

**Comment by sanket1729** [comment]
@Velnbur While the API is indeed awkward, you would need two allocations for script anyways. The third one can be prevented as follows:

ControlBlock is typically created at spend_time while the taptree is created at address creation time. TaptreeBuilder must consume the script to maintain the tree stucture, so the first clone feels unavoidable.  For the second clone, you can get around the burrow checker as
```rust
let script_ver_pair = (script, LeafVersion::Tapscript);
let control_block = spending_info.control_block(&script_ver_pair);
println!("{}", script.0.to_asm_string());
```

**Comment by Velnbur** [comment]
> TaptreeBuilder must consume the script to maintain the tree structure, so the first clone feels unavoidable.

From my point of view, I think it's possible to have `Cow<'script, Script>` inside `TaptreeBuilder` instead of `ScriptBuf`. This would require the introduction of a `'script` lifetime, but still will minimize the breaking changes as would work for both `ScriptBuf` and `Script`. 

Considering the new implementation as `TaptreeBuilderNew<'script>`, the old one should mostly be `type TaptreeBuilder = TaptreeBuilderNew<'static>` by doing:

1. Change all `script: ScriptBuf` parameters to `script: impl Into<Cow<'script, Script>>`;
2. Implement `From<ScriptBuf>` for `Cow<'script, Script>`.

But I'm not a maintainer, so I don't see the whole picture

**Comment by sanket1729** [comment]
> I think it's possible to have Cow<'script, Script> inside TaptreeBuilder instead of ScriptBuf.

Nice. I think can make `Cow<'script, Script>` work.

**Comment by Kixunil** [comment]
Please before jumping on a cow, consider using a generic that defaults to `ScriptBuf`. The callers could then pass in `Cow`, or `Rc`/`Arc` or whatever else makes sense for them. And it should be non-breaking since it's `ScriptBuf` by default.

**Comment by apoelstra** [comment]
I think we should index the tree by `TapLeafHash` and fix the API while we're at it to return an iterator of results in case there are multiple identical leaves (yes, this is crazy and pointless, but our current API responds to it by silently hiding one which is surprising behavior and pretty hard to detect when it's happened).

But those are just a couple changes I'd make. Really we need to step back and rethink this whole API. I'd like to do this in Miniscript first where our usage is more constrained to explore the design space.

I'm not worried about breaking changes with this API. Only a few people have tried to use it because it's so crazy and undiscoverable.

**Comment by Kixunil** [comment]
Yeah, I prefer hard-coded tap tree anyway for things I know at compile time.
